### PR TITLE
[lte] cellular fixes

### PR DESCRIPTION
**Pull request by technobly**
### Problem

There are several issues currently:
- u-blox SARA-R410M-02B firmware 05.06,A.02.00 has a memory housekeeping issue where after every 124 days it does a cleanup.  If the power is forced off while that cleanup is in progress, the modem can enter an unrecoverable state.  u-blox has advised on proper timeouts from power up and registration to mitigate this issue.  Only this particular version of u-blox firmware has this issue.
- u-blox SARA-R410M-02B firmware 05.08,A.02.04 changes the way some line endings are generated for multiple commands.  This breaks the AT command parser on Gen 2 devices.  Gen 3 is unaffected based on how it removes all line endings before parsing commands.
- u-blox SARA-R410M-02B firmware 05.08,A.02.04 requires the AT+UMNOPROF default setting to be changed to a specific carrier, or SIM Select.  We prefer SIM Select as it should allow 3rd party SIMs to have the correct settings applied.

### Solution

This PR solves the above issues by implementing a commit for each one:
- [gen 2] mitigates memory issue in u-blox firmware 05.06,A.02.00 [ch33464]
- [gen 2] fixes AT command parser issues with u-blox firmware 05.08,A.02.04 [ch32049]
- [gen 2] adds AT+UMNOPROF=1 (SIM Select) setting [ch29063]
- [gen 2] adds debugging for Pipe buffer with MDM_DEBUG_RX_PIPE & MDM_DEBUG_TX_PIPE defines
- [gen 3] mitigates memory issue in u-blox firmware 05.06,A.02.00 [ch33464]
- [gen 3] adds AT+UMNOPROF=1 (SIM Select) setting [ch29063]

### Steps to Test

- Run unit tests
- Run TEST=wiring/no_fixture
- Lots of manual testing on U260, G350, R410 modems with new and old u-blox firmware

### References

Closes [ch33464]
Closes [ch32049]
Closes [ch29063]

---

### Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed CLA ([Info here](https://github.com/spark/firmware/blob/develop/CONTRIBUTING.md))
- [x] Problem and Solution clearly stated
- [x] Run unit/integration/application tests on device
- [x] (N/A) Added documentation
- [ ] Added to CHANGELOG.md after merging (add links to docs and issues)

### CODE REVIEW

File: hal/src/boron/network/cellular_ncp_client.h [hunk 1]
```diff
@@ -49,6 +49,22 @@ class CellularNcpClientConfig: public NcpClientConfig {
     MeshNCPIdentifier ident_;
 };
 
+enum class UbloxSaraUMNOPROF {
```
Comment (avtolstoy) [inline hal/src/boron/network/cellular_ncp_client.h]: Nitpicking: prefer `UbloxSaraUmnoprof`.

File: hal/src/boron/network/sara_ncp_client.cpp [hunk 2]
```diff
@@ -790,15 +796,58 @@ int SaraNcpClient::initReady() {
...
+        resp = parser_.sendCommand("AT+UMNOPROF?");
+        bool reset = false;
+        int umnoprof = static_cast<int>(UbloxSaraUMNOPROF::NONE);
+        r = CHECK_PARSER(resp.scanf("+UMNOPROF: %d", &umnoprof));
         CHECK_PARSER_OK(resp.readResult());
-        if (selectAct != 7 || (r >= 2 && preferAct1 != 7) || (r >= 3 && preferAct2 != 7)) { // 7: LTE Cat M1
+        if (r == 1 && static_cast<UbloxSaraUMNOPROF>(umnoprof) == UbloxSaraUMNOPROF::SW_DEFAULT) {
             // This is a persistent setting
-            CHECK_PARSER_OK(parser_.execCommand("AT+URAT=7"));
+            auto respUmno = parser_.sendCommand(1000, "AT+UMNOPROF=%d\r\n", static_cast<int>(UbloxSaraUMNOPROF::SIM_SELECT));
```
Comment (avtolstoy) [inline hal/src/boron/network/sara_ncp_client.cpp]: `\r\n` is not needed.

File: hal/src/electron/modem/enums_hal.h [hunk 3]
```diff
@@ -98,6 +98,22 @@ typedef enum {
...
+    UBLOX_SARA_UMNOPROF_VERIZON          = 3,
+    UBLOX_SARA_UMNOPROF_TELSTRA          = 4,
+    UBLOX_SARA_UMNOPROF_TMOBILE          = 5,
+    UBLOX_SARA_UMNOPROF_CHINA_TELECOM    = 6,
+    UBLOX_SARA_UMNOPROF_SPRINT           = 8,
+    UBLOX_SARA_UMNOPROF_VODAFONE         = 19,
+    UBLOX_SARA_UMNOPROF_TELUS            = 21,
+    UBLOX_SARA_UMNOPROF_DEUTSCHE_TELEKOM = 31,
+    UBLOX_SARA_UMNOPROF_STANDARD_EUROPE  = 100,
+} UbloxSaraUMNOPROF;
```
Comment (avtolstoy) [inline hal/src/electron/modem/enums_hal.h]: Nitpicking: prefer `UbloxSaraUmnoprof`.

File: hal/src/boron/network/sara_ncp_client.cpp [hunk 4]
```diff
@@ -1271,4 +1336,26 @@ int SaraNcpClient::modemSetUartState(bool state) const {
...
+        now = millis();
+        // prefer to timeout 20s after registration if we are registered
+        if (registeredTime_) {
+            if (now - registeredTime_ >= 20000UL) {
+                break;
+            }
+        } else if (powerOnTime_ && now - powerOnTime_ >= 30000UL) {
+            break;
+        }
+        HAL_Delay_Milliseconds(1); // just wait
```
Comment (avtolstoy) [inline hal/src/boron/network/sara_ncp_client.cpp]: Let's increase this to something like `100`.

File: hal/src/electron/modem/mdm_hal.cpp [hunk 5]
```diff
@@ -951,16 +1007,47 @@ bool MDMParser::powerOff(void)
...
+                            now = HAL_Timer_Get_Milli_Seconds();
+                            // prefer to timeout 20s after registration if we are registered
+                            if (_timeRegistered) {
+                                if (now - _timeRegistered >= 20000UL) {
+                                    break;
+                                }
+                            } else if (_timePowerOn && now - _timePowerOn >= 30000UL) {
+                                break;
+                            }
+                            HAL_Delay_Milliseconds(1); // just wait
```
Comment (avtolstoy) [inline hal/src/electron/modem/mdm_hal.cpp]: Let's increase this to something like `100`.

File: hal/src/boron/network/sara_ncp_client.cpp (hunk 2 above)
Comment (technobly) [reply hal/src/boron/network/sara_ncp_client.cpp]: Such habit!  Done.

File: hal/src/electron/modem/enums_hal.h (hunk 3 above)
Comment (technobly) [reply hal/src/electron/modem/enums_hal.h]: Totes, no problem. Done.

File: hal/src/boron/network/sara_ncp_client.cpp (hunk 4 above)
Comment (technobly) [reply hal/src/boron/network/sara_ncp_client.cpp]: Ok, done.

File: hal/src/electron/modem/mdm_hal.cpp (hunk 5 above)
Comment (technobly) [reply hal/src/electron/modem/mdm_hal.cpp]: Ok, done.

**Pull request merged**